torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,257 @@
+            maybe_inplace_op = maybe_get_inplace_op(node.target)
+            if maybe_inplace_op is not None:
+                # This is a proxy check for ensuring that the first argument is "tensor-like"
+                # (This should be the case for all ops with inplace variants in ATen,
+                # although we technically don't have guarantees for custom ops).
+                assert len(node.target._schema.arguments) > 0
+                assert 'Tensor' in str(node.target._schema.arguments[0].type)
+
+                # Step 2: Check to see if the input to the op is re-used later in the graph.
+                self_arg = node.args[0]
+                self_arg_name = self_arg.name
+
+                input_alias_names = set()
+                for inpt_name in input_names:
+                    input_alias_names |= alias_map[inpt_name]
+
+                input_aliases_of_current_tensor = [n for n in input_alias_names if n in alias_map[self_arg_name]]
+
+                # In general, we can't re-inplace an op if input was
+                # an (alias to an) input to the graph.
+                if len(input_aliases_of_current_tensor) > 0:
+                    # However, if that input is overwritten by a copy_() later in the graph, then
+                    # it's fair game to re-inplace.
+                    if not any(last_node_where_input_alias_is_overwritten[inpt_alias_idx] > idx for inpt_alias_idx in input_aliases_of_current_tensor):
+                        continue
+
+                # For the current self_tensor, we want to know:
+                # What are all of the nodes in the graph that use self_tensor, or any of its aliases, as an input?
+                self_arg_aliases = alias_map[self_arg_name]
+                nodes_used = concatMap(lambda alias_name: names_to_used_nodes[alias_name], self_arg_aliases)
+                # for every node that uses the current variable, we've tracked what their index is
+                # in the graph. If there aren't any nodes with a higher index that use the current
+                # argument, that means it isn't re-used anywhere.
+                if not any(node_idx > idx for node_idx, _ in nodes_used):
+                    # Step 3: replace the current out-of-place op with its inplace variant.
+                    node.target = maybe_inplace_op
+
+                    # Step 4:
+                    # Now that we've replaced b = a.foo() with a.foo_(),
+                    # We need to replace any later usages of "b" with "a"
+                    #
+                    # A lot of the extra below is because we have several data structures
+                    # That we need to update when we change the nodes' arguments:
+                    # - names_to_used_nodes (some nodes now take in different variables as arguments)
+                    # - alias_map (updating a view nodes' arguments will change with variables alias)
```

Comment:
This part definitely feels like it would be easier if you were rebuilding the graph instead of mutating. You'd maintain a mapping env and apply it as you progressed further